### ISSUE TITLE
6.4.2 brought memory leak

### ISSUE THREAD

**Issue by baotongw** [log4js-node/log4js-node#1216]
After upgrade log4js to 6.4.2, I noticed the memory leak happened. My app have multi process that served by pm2, and each process share the same log file. when the log rolling zip on the hour, one of processes(or part of them) will lost the write ability, and the data just stay in the memory which made it grow up until reach the memory limitation. 

My node version: 14.16.0

**Comment by peteriman** [comment]
1. What was your prior version that was working?
2.  Would you mind trying the following versions and reporting the findings?
    - `log4js@6.3.0`
    - `log4js@6.4.0`
    - `log4js@6.4.3`

**Comment by peteriman** [comment]
1. Are you on Windows or Linux?
2. What was your prior version that was working?
3.  Would you mind trying the following versions and reporting the findings?
    - `log4js@6.3.0`
    - `log4js@6.4.0`
    - `log4js@6.4.3`

**Comment by baotongw** [comment]
It's Linux centos, prior version is 4.5.1. 

ok, I will downgrade to 6.3.0 and see what going next

thanks

**Comment by peteriman** [comment]
Yes please, that will help my investigation greatly.

**Comment by baotongw** [comment]
I downgrade to 6.3.0, memory leak gone, but got another similar problems. when multi process write log to one file, some low write frequency process data was dropped. I wrote a test code that can repro the issue.

`
// logger.js
const log4js = require('log4js')

log4js.configure({
  appenders: {
    access: {
      type: 'dateFile',
      filename: './logs/access.log',
      pattern: '-yyyy-MM-dd-hh:mm.log',
      alwaysIncludePattern: true,
      category: 'access',
      compress: true,
      numBackups: 15 * 24,
    },
    error: {
      type: 'dateFile',
      filename: './logs/error.log',
      pattern: 'yyyy-MM-dd-hh:mm.log',
      alwaysIncludePattern: true,
      category: 'error',
      compress: true,
      numBackups: 15 * 24,
    },
  },
  categories: {
    default: {
      appenders: ['access'],
      level: 'info',
    },
    error: {
      appenders: ['error'],
      level: 'error',
    }
  }
})

const infoLoger = log4js.getLogger();
const errorLoger = log4js.getLogger('error');

function info(...params) {
  infoLoger.info(...params);
}

function error(...params) {
  errorLoger.error(...params);
}

module.exports = {
  info,
  error,
}

// task.js
const logger = require('./logger');

const pid = process.pid;

function getRandom() {
    return Math.floor(Math.random() * 1000);
}

function getMessage(val) {    
    const msg = `I am ${pid}, the data is: ${val}`;
    return msg;
}

function writeLog() {
    const millesecond = getRandom();
    const msg = getMessage(millesecond);
    logger.info(msg);
}

const random = getRandom();

setInterval(() => writeLog(), random)

logger.info(`Subprocess ${pid} start`);

module.exports = writeLog;

// index.js
const http = require('http');
const { fork } = require('child_process');
const path = require('path');
const logger = require('./logger');

const taskPath = path.resolve(__dirname, './task.js');

const pidCount = 6;

const pids = {};

for (let i = 0; i < pidCount; i++) {
  (function () {
    let pro = fork(taskPath);

    pids[pro.pid] = pro.pid;
    pro.on('exit', () => { 
      pids[pro.pid] = null;

      logger.info(`Subprocess:${pro.pid} exit`);
    })
  })()
}

const server = http.createServer((req, res) => {
  res.writeHead(200, { 'Content-Type': 'application/json' });
  const resp = JSON.stringify(pids);
  res.end(resp);
});

logger.info('[Server start on port 8000.]');

server.listen(8000);
`

I fork 6 process, in the first hour log file, all process wrote success, and in the next hour, 4 process data was dropped.

// first hour log summary
pi@raspberry-bt:~/workspace/middleware-test/log4js/logs $ zcat access.log.-2022-03-30-23.log | awk '{print $7}' | sort | uniq -c
  12086 1629,
   2883 1630,
   4342 1636,
   2585 1637,
   3933 1639,
  48274 1645,

// second hour log summary
pi@raspberry-bt:~/workspace/middleware-test/log4js/logs $ zcat access.log.-2022-03-31-00.log | awk '{print $7}' | sort | uniq -c
   4872 1630,
  81584 1645,

as you see only two process log alive. in the 6.4.2, the issue still there and the memory keep growing that make leak. This should be something with buffer, I remember the memory up to 4GB on the production which I set the max old space to it.

PS: I run the test code on the linux OS with raspberry.

**Comment by peteriman** [comment]
1. How do you detect the memory leak? 
    So that I can try to investigate with a consistent approach.
    
2. I have summarised your findings in the table below.
   Would you mind helping me to narrow down the memory-leak-causing version by completing the table?
   
   |`log4js` Version|Memory Leak|Remarks|
   |-|-|-|
   |`6.3.0`|❌|last known working version without memory leak|
   |`6.4.0`|(test this second)|to test if caused by this version|
   |`6.4.1`|(skip if `6.4.0` tested to have memory leak)|to test if caused by this version|
   |`6.4.2`|✔️|has memory leak|
   |`6.4.3`|(skip)|not required|
   |`6.4.4`|(test this first)|to test if already fixed|

3. For the `child_process` dying issue:

    1. Did you get any `Subprocess:${pro.pid} exit` log messages?
    
    2. You might want to handle `pro.on('error')` as the [`'exit'` event may or may not fire after an error has occurred](https://nodejs.org/api/child_process.html#event-error). 
        Print out the `error` so that we can know what's going on.

**Comment by peteriman** [comment]
I ran your codes and I have this in my console when the drop of forked processes occurs (triggered during file rolling).

```
(node:40546) UnhandledPromiseRejectionWarning: Error: ENOENT: no such file or directory, open 'logs/access.log.-2022-04-19-11.log'
(Use `node --trace-warnings ...` to show where the warning was created)
(node:40546) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:40546) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

**Comment by peteriman** [comment]
I might have found the problem.

![image](https://user-images.githubusercontent.com/29530649/164887490-3e1cddf2-1f89-433f-8770-e35778d6f270.png)
_(src: https://github.com/log4js-node/streamroller/blob/master/lib/moveAndMaybeCompressFile.js#L27-L47)_

While line 27 `await fs.pathExists(sourceFilePath)` is `true`, line 35 `fs.createReadStream(sourceFilePath)` can still throw error due to concurrency issue that a forked process has deleted in `sourceFilePath` when the current logic is running between the 2 lines. 

When that happens, an `error` event is emitted for `ReadStream` which is not handled. 

 `WriteStream` will emit `finish` event regardless, causing entry into line 38 `.on("finish", () => {`. It attempts to unlink the `sourceFilePath` which fails and fallbacks to truncate the `sourceFilePath` which still fails as the file does not exist. Finally, rejection the promise.

The promise is declared to the await at line 34 `await new Promise((resolve, reject) => {`, yet there is no handler for promise rejection.

**Comment by peteriman** [comment]
@baotongw 

You can try replacing your `node_modules/streamroller/lib/moveAndMaybeCompressFile.js` with the new [moveAndMaybeCompressFile.js](https://github.com/log4js-node/streamroller/blob/2d4368d76bb2845c0cababa71e0d68f7b9049fea/lib/moveAndMaybeCompressFile.js).

I will make a release in the next few days.

**Comment by baotongw** [comment]
great. I will test it in my demo tonight, and give feedback tomorrow.

**Comment by peteriman** [comment]
> so I don't think the finish is emitted, the next pipe should not be executed. I test it with below code, the console in the finish event is not print.

You are right but it doesn't matter as the Promise rejection from line 35 `fs.createReadStream(source)` is unhandled anyway.

**Comment by peteriman** [comment]
Fixed in `streamroller@3.0.8` - [changelog](https://github.com/log4js-node/streamroller/blob/master/CHANGELOG.md).
Fixed in `log4js@6.4.6` (which has dependency bumped to `streamroller@3.0.8`).

**Comment by peteriman** [comment]
Fixed in `streamroller@3.0.8` - [changelog](https://github.com/log4js-node/streamroller/blob/master/CHANGELOG.md).
Fixed in `log4js@6.4.6` (which has dependency bumped to `streamroller@3.0.8`).

@baotongw Please let me know if there are still issues after upgrading.

**Comment by peteriman** [comment]
The underlying cause is inherent from https://github.com/log4js-node/streamroller/commit/955e96ef53a356910193663126fc974feb67181f ([line 31 - 43](https://github.com/log4js-node/streamroller/blob/955e96ef53a356910193663126fc974feb67181f/lib/RollingFileWriteStream.js#L31-L43)) which is released in `streamroller@2.0.0`.

Hence, `streamroller@2.0.0 - 3.0.7` is affected by the unhandled promise rejection at [line 43](https://github.com/log4js-node/streamroller/blob/955e96ef53a356910193663126fc974feb67181f/lib/RollingFileWriteStream.js#L31-L43).
The affected versions of log4js that uses those streamroller versions will be `log4js@5.0.0 - 6.4.5`.

I will change the issue title for clarity.

**Comment by peteriman** [comment]
Fixed in `streamroller@3.0.8` - [changelog](https://github.com/log4js-node/streamroller/blob/master/CHANGELOG.md).
Fixed in `log4js@6.4.6` (which has dependency bumped to `streamroller@3.0.8`).

@baotongw Please let me know if there are still issues after upgrading.

**Comment by baotongw** [comment]
I upgraded it on 04-26 and it looks fine so far. all log data wrote and memory keep stable. since the memory growing slowly I will keep an eye on it.
